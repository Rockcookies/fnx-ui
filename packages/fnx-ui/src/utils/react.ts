import { Children, ReactElement } from 'react';
import { isFragment } from 'react-is';

export const toElementArray = (
	children: React.ReactNode,
	option?: { keepEmpty?: boolean },
): ReactElement[] => {
	const keepEmpty = option && option.keepEmpty;
	let ret: React.ReactElement[] = [];

	Children.forEach(children, (child: any) => {
		if ((child === undefined || child === null) && !keepEmpty) {
			return;
		}

		if (Array.isArray(child)) {
			ret = ret.concat(toElementArray(child));
		} else if (isFragment(child) && child.props) {
			ret = ret.concat(toElementArray(child.props.children, option));
		} else {
			ret.push(child);
		}
	});

	return ret;
};
