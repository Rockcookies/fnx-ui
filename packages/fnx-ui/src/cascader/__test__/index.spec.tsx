import { fireEvent, render, waitFor } from '@testing-library/react';
import React from 'react';
import Cascader from '../';
import { Icon } from '../../';
import TestsDOM from '../../../test/dom';
import TestsEvent from '../../../test/events';

describe('<Cascader/>', () => {
	const areaData = [
		{
			label: 'China',
			children: [
				{
					label: 'Beijing',
					children: [{ label: 'Haidian' }, { label: 'Chaoyang' }],
				},
				{
					label: 'Shanghai',
					children: [{ label: 'Zhabei' }, { label: 'Jingan' }],
				},
			],
		},
		{
			label: 'United State',
			children: [
				{
					label: 'New York',
					children: [{ label: 'Manhattan' }, { label: 'Brooklyn' }],
				},
				{
					label: 'California',
					children: [
						{ label: 'Los Angeles' },
						{ label: 'San Francisco' },
					],
				},
			],
		},
	];

	const mockRect = (el: Element, rect: Partial<DOMRect>) => {
		Object.defineProperty(el, 'getBoundingClientRect', {
			configurable: true,
			get: () => {
				return () => ({
					top: 0,
					left: 0,
					right: 0,
					bottom: 0,
					width: 0,
					height: 0,
					...rect,
				});
			},
		});
	};

	beforeEach(() => {
		jest.useFakeTimers();
	});

	afterEach(() => {
		jest.useRealTimers();
	});

	it('defaultValue and dataNames props', () => {
		const { container } = render(
			<Cascader
				data={areaData}
				dataNames={{ value: 'label' }}
				defaultValue={['China', 'Beijing', 'Haidian']}
			></Cascader>,
		);

		const tabs = container.querySelectorAll('.fnx-tabs__tab');

		expect(
			tabs[2].classList.contains('fnx-tabs__tab--active'),
		).toBeTruthy();

		expect(container).toMatchSnapshot();
	});

	it('onConfirm、 onChange should trigger', () => {
		const onConfirm = jest.fn();
		const onChange = jest.fn();

		const { container } = render(
			<Cascader
				data={areaData}
				dataNames={{ value: 'label' }}
				onConfirm={(v) => onConfirm(v)}
				onChange={(v) => onChange(v)}
			></Cascader>,
		);

		fireEvent.click(
			container
				.querySelectorAll('.fnx-swipe-item')[0]
				.querySelectorAll('.fnx-cascader__option')[0],
		);
		expect(onChange).toHaveBeenCalledWith(['China']);

		fireEvent.click(
			container
				.querySelectorAll('.fnx-swipe-item')[1]
				.querySelectorAll('.fnx-cascader__option')[0],
		);
		expect(onChange).toHaveBeenCalledWith(['China', 'Beijing']);

		fireEvent.click(
			container
				.querySelectorAll('.fnx-swipe-item')[2]
				.querySelectorAll('.fnx-cascader__option')[0],
		);
		expect(onChange).toHaveBeenCalledWith(['China', 'Beijing', 'Haidian']);
		expect(onConfirm).toHaveBeenCalledWith(['China', 'Beijing', 'Haidian']);
	});

	it('title、placeholder props', () => {
		const { container } = render(
			<Cascader
				data={areaData}
				dataNames={{ value: 'label' }}
				title="please choose area"
				placeholder="placeholder"
			></Cascader>,
		);

		expect(
			TestsDOM.mustQuerySelector(container, '.fnx-cascader__title')
				.innerHTML,
		).toEqual('please choose area');

		expect(
			TestsDOM.mustQuerySelector(
				container,
				'.fnx-cascader__tab-title--unselected',
			).innerHTML,
		).toEqual('placeholder');
	});

	it('onCancel、closeIcon props', () => {
		const onCancel = jest.fn();
		const { container } = render(
			<Cascader
				data={areaData}
				dataNames={{ value: 'label' }}
				closeIcon={<Icon name="close-o" />}
				onCancel={onCancel}
			></Cascader>,
		);

		fireEvent.click(
			TestsDOM.mustQuerySelector(container, '.fnx-cascader__close-icon'),
		);

		expect(onCancel).toHaveBeenCalled();
		expect(container).toMatchSnapshot();
	});

	it('activeColor props', () => {
		const { container } = render(
			<Cascader
				data={areaData}
				dataNames={{ value: 'label' }}
				defaultValue={[areaData[0].label]}
				activeColor="red"
			></Cascader>,
		);

		const selected = TestsDOM.mustQuerySelector(
			container,
			'.fnx-cascader__option--selected',
		);

		expect(getComputedStyle(selected).color).toEqual('red');
	});

	it('swipeable props', async () => {
		const { container } = render(
			<Cascader
				data={areaData}
				dataNames={{ value: 'label' }}
				swipeable={true}
			></Cascader>,
		);

		const swiper = TestsDOM.mustQuerySelector(
			container,
			'.fnx-swipe__wrapper',
		);

		mockRect(TestsDOM.mustQuerySelector(container, '.fnx-swipe'), {
			width: 300,
			height: 150,
		});

		const swipeItem = container.querySelectorAll('.fnx-swipe-item');

		fireEvent.click(
			swipeItem[0].querySelectorAll('.fnx-cascader__option')[0],
		);

		const tabs = container.querySelectorAll('.fnx-tabs__tab');

		expect(
			tabs[1].classList.contains('fnx-tabs__tab--active'),
		).toBeTruthy();

		await TestsEvent.triggerDrag(swiper, [100, 0]);

		await waitFor(() => {
			expect(
				tabs[0].classList.contains('fnx-tabs__tab--active'),
			).toBeTruthy();
		});
	});

	it('onLoadData props', async () => {
		const _data: any[] = [{ label: 'China' }, { label: 'American' }];

		const onLoadData = jest.fn(() => {
			_data[0] = { label: 'China', loading: true };
			rerender(
				<Cascader
					dataNames={{ value: 'label' }}
					onLoadData={onLoadData}
					data={[..._data]}
				></Cascader>,
			);
		});

		const { container, rerender } = render(
			<Cascader
				data={_data}
				dataNames={{ value: 'label' }}
				onLoadData={onLoadData}
			></Cascader>,
		);

		const swipeItem = container.querySelectorAll('.fnx-swipe-item');

		fireEvent.click(
			swipeItem[0].querySelectorAll('.fnx-cascader__option')[0],
		);

		expect(
			container.querySelector('.fnx-cascader__loading'),
		).not.toBeNull();
	});

	it('select set disabled ', async () => {
		const _data: any[] = [...areaData];
		_data[0].children[0].disabled = true;
		const onChange = jest.fn();

		const { container } = render(
			<Cascader
				data={_data}
				dataNames={{ value: 'label' }}
				defaultValue={['China', 'Beijing', 'Haidian']}
				onChange={onChange}
			></Cascader>,
		);

		const tabs = container.querySelectorAll('.fnx-tabs__tab');
		expect(
			tabs[0].classList.contains('fnx-tabs__tab--active'),
		).toBeTruthy();

		fireEvent.click(container.querySelectorAll('.fnx-tabs__tab')[1]);
		fireEvent.click(
			container
				.querySelectorAll('.fnx-swipe-item')[1]
				.querySelectorAll('.fnx-cascader__option')[0],
		);

		expect(onChange).not.toHaveBeenCalled();
	});
});
