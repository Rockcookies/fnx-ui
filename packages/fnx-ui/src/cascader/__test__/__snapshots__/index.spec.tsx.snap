// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Cascader/> defaultValue and dataNames props 1`] = `
<div>
  <div
    class="fnx-cascader"
  >
    <div
      class="fnx-cascader__header"
    >
      <div
        class="fnx-cascader__title"
      />
      <span
        class="fnx-cascader__close-icon"
      >
        <span
          class="fnx-icon"
        >
          <svg
            aria-hidden="true"
            fill="currentColor"
            focusable="false"
            height="1em"
            viewBox="0 0 1024 1024"
            width="1em"
          >
            <path
              d="M573.592576 513.252352l261.472256 261.472256c16.662528 16.662528 16.662528 43.677696 0 60.340224s-43.677696 16.662528-60.340224 0L513.252352 573.592576 251.78112 835.064832c-16.662528 16.662528-43.677696 16.662528-60.3392 0-16.662528-16.662528-16.662528-43.677696 0-60.340224l261.472256-261.472256L191.440896 251.78112c-16.662528-16.662528-16.662528-43.677696 0-60.3392 16.661504-16.662528 43.676672-16.662528 60.3392 0l261.472256 261.472256 261.472256-261.472256c16.662528-16.662528 43.677696-16.662528 60.340224 0 16.662528 16.661504 16.662528 43.676672 0 60.3392L573.592576 513.252352z"
            />
          </svg>
        </span>
      </span>
    </div>
    <div
      class="fnx-tabs fnx-cascader__tabs"
    >
      <div
        class="fnx-tabs__header"
      >
        <div
          class="fnx-tabs__nav"
        >
          <div
            class="fnx-tabs__tab-list fnx-tabs__tab-list--scrollable"
            role="tablist"
          >
            <div
              aria-selected="false"
              class="fnx-tabs__tab"
              role="tab"
            >
              <div
                class="fnx-tabs__tab-title"
              >
                <span
                  class="fnx-cascader__tab-title"
                >
                  China
                </span>
              </div>
            </div>
            <div
              aria-selected="false"
              class="fnx-tabs__tab"
              role="tab"
            >
              <div
                class="fnx-tabs__tab-title"
              >
                <span
                  class="fnx-cascader__tab-title"
                >
                  Beijing
                </span>
              </div>
            </div>
            <div
              aria-selected="true"
              class="fnx-tabs__tab fnx-tabs__tab--active"
              role="tab"
            >
              <div
                class="fnx-tabs__tab-title"
              >
                <span
                  class="fnx-cascader__tab-title"
                >
                  Haidian
                </span>
              </div>
            </div>
            <div
              class="fnx-tabs__track"
              style="left: 0px; transition-duration: 300ms;"
            />
          </div>
        </div>
      </div>
      <div
        class="fnx-tabs__content fnx-tabs__content--animated"
      >
        <div
          class="fnx-swipe fnx-tabs__swipe"
        >
          <div
            class="fnx-swipe__wrapper"
            style="transform: translateX(0px);"
          >
            <div
              aria-hidden="true"
              class="fnx-swipe-item fnx-tabs__swipe-item fnx-tabs__swipe-item--inactive"
              data-fnx-tab-panel-key="0"
              role="tabpanel"
            >
              <div
                class="fnx-tabs__panel"
              />
            </div>
            <div
              aria-hidden="true"
              class="fnx-swipe-item fnx-tabs__swipe-item fnx-tabs__swipe-item--inactive"
              data-fnx-tab-panel-key="1"
              role="tabpanel"
            >
              <div
                class="fnx-tabs__panel"
              />
            </div>
            <div
              aria-hidden="false"
              class="fnx-swipe-item fnx-tabs__swipe-item"
              data-fnx-tab-panel-key="2"
              role="tabpanel"
            >
              <div
                class="fnx-tabs__panel"
              >
                <div
                  class="fnx-cascader__options-container"
                >
                  <ul
                    class="fnx-cascader__options"
                  >
                    <li
                      class="fnx-cascader__option fnx-cascader__option--selected"
                    >
                      <span
                        class="fnx-cascader__label"
                      >
                        Haidian
                      </span>
                      <span
                        class="fnx-icon fnx-cascader__selected-icon"
                      >
                        <svg
                          aria-hidden="true"
                          fill="currentColor"
                          focusable="false"
                          height="1em"
                          viewBox="0 0 1024 1024"
                          width="1em"
                        >
                          <path
                            d="M122.28096 536.623104c-9.940992-9.925632-11.548672-25.360384-2.78528-36.407296l20.487168-25.828352c8.397824-10.58816 24.108032-13.246464 35.211264-5.835776l177.3312 118.35904c9.353216 6.243328 25.452544 5.430272 34.185216-1.654784l468.5824-380.16c10.532864-8.54528 27.030528-7.817216 36.261888 1.400832l11.542528 11.52512c10.04544 10.03008 9.314304 25.951232-1.215488 36.465664l-502.92736 502.183936c-15.64672 15.624192-41.337856 14.94016-57.445376-1.142784l-219.22816-218.9056z"
                          />
                        </svg>
                      </span>
                    </li>
                    <li
                      class="fnx-cascader__option"
                    >
                      <span
                        class="fnx-cascader__label"
                      >
                        Chaoyang
                      </span>
                    </li>
                  </ul>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`<Cascader/> onCancel„ÄÅcloseIcon props 1`] = `
<div>
  <div
    class="fnx-cascader"
  >
    <div
      class="fnx-cascader__header"
    >
      <div
        class="fnx-cascader__title"
      />
      <span
        class="fnx-cascader__close-icon"
      >
        <span
          class="fnx-icon"
        >
          <svg
            aria-hidden="true"
            fill="currentColor"
            focusable="false"
            height="1em"
            viewBox="0 0 1024 1024"
            width="1em"
          >
            <path
              d="M511.65696 471.998464l140.791808-140.792832c11.108352-11.108352 29.118464-11.108352 40.226816 0 11.108352 11.108352 11.108352 29.118464 0 40.226816L551.882752 512.22528l140.792832 140.792832c11.108352 11.108352 11.108352 29.118464 0 40.226816-11.108352 11.108352-29.118464 11.108352-40.226816 0L511.65696 552.452096 370.864128 693.244928c-11.108352 11.108352-29.118464 11.108352-40.226816 0-11.108352-11.108352-11.108352-29.118464 0-40.226816L471.42912 512.22528 330.637312 371.432448c-11.108352-11.108352-11.108352-29.118464 0-40.226816 11.108352-11.108352 29.118464-11.108352 40.226816 0l140.792832 140.792832z"
            />
            <path
              d="M512 910.222336c219.931648 0 398.222336-178.290688 398.222336-398.222336 0-219.931648-178.290688-398.222336-398.222336-398.222336-219.931648 0-398.222336 178.290688-398.222336 398.222336 0 219.931648 178.290688 398.222336 398.222336 398.222336z m0 56.88832C260.64896 967.110656 56.889344 763.35104 56.889344 512S260.64896 56.889344 512 56.889344 967.110656 260.64896 967.110656 512 763.35104 967.110656 512 967.110656z"
            />
          </svg>
        </span>
      </span>
    </div>
    <div
      class="fnx-tabs fnx-cascader__tabs"
    >
      <div
        class="fnx-tabs__header"
      >
        <div
          class="fnx-tabs__nav"
        >
          <div
            class="fnx-tabs__tab-list fnx-tabs__tab-list--scrollable"
            role="tablist"
          >
            <div
              aria-selected="true"
              class="fnx-tabs__tab fnx-tabs__tab--active"
              role="tab"
            >
              <div
                class="fnx-tabs__tab-title"
              >
                <span
                  class="fnx-cascader__tab-title fnx-cascader__tab-title--unselected"
                >
                  Select
                </span>
              </div>
            </div>
            <div
              class="fnx-tabs__track"
              style="left: 0px; transition-duration: 300ms;"
            />
          </div>
        </div>
      </div>
      <div
        class="fnx-tabs__content fnx-tabs__content--animated"
      >
        <div
          class="fnx-swipe fnx-tabs__swipe"
        >
          <div
            class="fnx-swipe__wrapper"
            style="transform: translateX(0px);"
          >
            <div
              aria-hidden="false"
              class="fnx-swipe-item fnx-tabs__swipe-item"
              data-fnx-tab-panel-key="0"
              role="tabpanel"
            >
              <div
                class="fnx-tabs__panel"
              >
                <div
                  class="fnx-cascader__options-container"
                >
                  <ul
                    class="fnx-cascader__options"
                  >
                    <li
                      class="fnx-cascader__option"
                    >
                      <span
                        class="fnx-cascader__label"
                      >
                        China
                      </span>
                    </li>
                    <li
                      class="fnx-cascader__option"
                    >
                      <span
                        class="fnx-cascader__label"
                      >
                        United State
                      </span>
                    </li>
                  </ul>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
